Caddy-tris: ERR_SSL_VERSION_INTERFERENCE

I've built caddy with the tris crypto library of go and thus tried out to use tls1.3 with caddy.

I added/modified three lines to the caddytls/config.go: `"TLS1-3": TLS_CHACHA20_POLY1305_SHA256` as an additional cipher and to only use this one for TLS1.3 negotiation and `"tls1.3": tls.VersionTLS13` for the available TLS versions.

Additionally I stated in the same file, that the max version is tls1.3: 

```
if config.ProtocolMaxVersion == 0 {
                config.ProtocolMaxVersion = tls.VersionTLS13
        }
```
But still [dev.ssllabs.com/ssltest](https://dev.ssllabs.com/ssltest/analyze.html?d=h-neef.de&s=185.101.92.192&latest) says, that my server negotiates TLS1.3 with a blacklisted HTTP/2 ciphersuite, which is why probably chrome throws the error mentioned in the title.

Any ideas?

Hi,

just tested to compile it, but I get the following error:

```
[319/411] Linking CXX executable crypto/crypto_test
[320/411] Building CXX object ssl/CMakeFiles/ssl.dir/bio_ssl.cc.o
[321/411] Building CXX object ssl/CMakeFiles/ssl.dir/d1_both.cc.o
[322/411] Building CXX object ssl/CMakeFiles/ssl_test.dir/ssl_test.cc.o
FAILED: ssl/CMakeFiles/ssl_test.dir/ssl_test.cc.o
/usr/bin/c++  -DBORINGSSL_IMPLEMENTATION -I../third_party/googletest/include -I.                                                                        ./ssl/../include -Werror -Wformat=2 -Wsign-compare -Wmissing-field-initializers                                                                         -Wwrite-strings -Wall -ggdb -fvisibility=hidden -fno-common -Wno-free-nonheap-ob                                                                        ject -Wmissing-declarations -std=c++11 -fno-exceptions -fno-rtti -Wshadow -MD -M                                                                        T ssl/CMakeFiles/ssl_test.dir/ssl_test.cc.o -MF ssl/CMakeFiles/ssl_test.dir/ssl_                                                                        test.cc.o.d -o ssl/CMakeFiles/ssl_test.dir/ssl_test.cc.o -c ../ssl/ssl_test.cc
c++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <http://gcc.gnu.org/bugs.html> for instructions.
[323/411] Building CXX object ssl/CMakeFiles/ssl.dir/d1_lib.cc.o
ninja: build stopped: subcommand failed.
The command '/bin/sh -c cd boringssl && ninja -C build' returned a non-zero code                                                                        : 1
_dev/Makefile:59: recipe for target 'build-test-boring' failed
make: *** [build-test-boring] Error 1
root@h-neef:~/tls-tris#
```

--- a/_dev/interop_test_runner
+++ b/_dev/interop_test_runner
@@ -15,6 +15,8 @@ RE_PATTERN_HELLO_TLS_13_RESUME   = "Hello TLS 1.3 \[resumed\] _o/"
 RE_PATTERN_HELLO_0RTT            = "^.*Hello TLS 1.3 .*\[resumed\] \[0-RTT\] _o/$"
 # Checks if 0-RTT was used and confirmed
 RE_PATTERN_HELLO_0RTT_CONFIRMED  = "^.*Hello TLS 1.3 .*\[resumed\] \[0-RTT confirmed\] _o/$"
+# ALPN
+RE_PATTERN_ALPN = "ALPN protocol: npn_proto$"
 
 class Docker(object):
     ''' Utility class used for starting/stoping servers and clients during tests'''
@@ -205,12 +207,20 @@ class InteropClient(object):
 # Actual test definition
 
 # TRIS as a server
-class InteropServer_BoringSSL(
-        InteropServer,
-        ServerNominalMixin,
-        ServerClientAuthMixin,
-        unittest.TestCase
-    ): CLIENT_NAME = "tls-tris:boring"
+class InteropServer_BoringSSL(InteropServer, ServerNominalMixin, ServerClientAuthMixin, unittest.TestCase):
+
+    CLIENT_NAME = "tls-tris:boring"
+
+    def test_ALPN(self):
+        '''
+        Checks wether ALPN is sent back by tris server in EncryptedExtensions in case of TLS 1.3. The
+        ALPN protocol is set to 'npn_proto', which is hardcoded in TRIS test server.
+        '''
+        res = self.d.run_client(self.CLIENT_NAME, self.server_ip+":1443 "+'-alpn-protos npn_proto -debug')
+        print(res[1])
+        self.assertEqual(res[0], 0)
+        self.assertIsNotNone(re.search(RE_PATTERN_ALPN, res[1], re.MULTILINE))
+
 
 # PicoTLS doesn't seem to implement draft-23 correctly. It will
 # be enabled when draft-28 is implemented.

--- a/_dev/tris-localserver/server.go
+++ b/_dev/tris-localserver/server.go
@@ -72,6 +72,7 @@ func (s *server) start() {
 	s.TLS.ClientCAs = x509.NewCertPool()
 	s.TLS.ClientCAs.AppendCertsFromPEM([]byte(rsaCa_client))
 	s.TLS.Accept0RTTData = ((s.ZeroRTT & ZeroRTT_Accept) == ZeroRTT_Accept)
+	s.TLS.NextProtos = []string{"npn_proto"}
 
 	httpServer := &http.Server{
 		Addr:      s.Address,

--- a/handshake_server.go
+++ b/handshake_server.go
@@ -277,10 +277,10 @@ Curves:
 
 	if len(hs.clientHello.alpnProtocols) > 0 {
 		if selectedProto, fallback := mutualProtocol(hs.clientHello.alpnProtocols, c.config.NextProtos); !fallback {
-			if hs.hello != nil {
-				hs.hello.alpnProtocol = selectedProto
-			} else {
+			if hs.hello13Enc != nil {
 				hs.hello13Enc.alpnProtocol = selectedProto
+			} else {
+				hs.hello.alpnProtocol = selectedProto
 			}
 			c.clientProtocol = selectedProto
 		}

